File: ai_sessions/providers/codex.py
from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path
from typing import Any

from ..config import yolo_default
from ..sessions import file_times, read_jsonl
from .base import RunResult, SessionDetail, SessionMessage, SessionSummary


def _codex_home() -> Path:
    return Path(os.environ.get("CODEX_HOME") or (Path.home() / ".codex"))


def _sessions_dir() -> Path:
    return _codex_home() / "sessions"


def _derive_id(path: Path) -> str:
    base = path.stem
    idx = base.rfind("-")
    return base[idx + 1 :] if idx > 0 else base


def _flatten(content: Any) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        out: list[str] = []
        for c in content:
            if isinstance(c, dict):
                out.append(c.get("text") or str(c))
            else:
                out.append(str(c))
        return "\n".join(out)
    return "" if content is None else str(content)


def _codex_command() -> list[str]:
    """Resolve the `codex app-server` command, handling Windows .cmd shims."""
    override = os.environ.get("CODEX_BIN")
    if override:
        return [override, "app-server"]
    # On Windows, npm-installed `codex` is `codex.cmd`; subprocess won't auto-resolve.
    candidates = ["codex.cmd", "codex.exe", "codex"] if sys.platform == "win32" else ["codex"]
    for name in candidates:
        found = shutil.which(name)
        if found:
            return [found, "app-server"]
    return ["codex", "app-server"]


def _build_thread_config(yolo: bool, cwd: str | None) -> Any:
    """Build a ThreadConfig with YOLO and cwd applied. Returns None if no overrides."""
    from codex_app_server_sdk import ThreadConfig

    kwargs: dict[str, Any] = {}
    if cwd:
        kwargs["cwd"] = cwd
    if yolo:
        kwargs["sandbox"] = "danger-full-access"
        kwargs["approval_policy"] = "never"
    if not kwargs:
        return None
    return ThreadConfig(**kwargs)


class CodexProvider:
    name = "codex"

    async def is_available(self) -> bool:
        return _sessions_dir().exists()

    async def list_sessions(self) -> list[SessionSummary]:
        root = _sessions_dir()
        if not root.exists():
            return []
        out: list[SessionSummary] = []
        for f in root.rglob("*.jsonl"):
            t = file_times(f)
            out.append(
                SessionSummary(
                    id=_derive_id(f),
                    provider="codex",
                    path=str(f),
                    createdAt=t["createdAt"],
                    updatedAt=t["updatedAt"],
                )
            )
        out.sort(key=lambda s: s.updatedAt or "", reverse=True)
        return out

    async def get_session(self, id: str) -> SessionDetail:
        root = _sessions_dir()
        match: Path | None = None
        for f in root.rglob("*.jsonl"):
            if _derive_id(f) == id or f.stem == id:
                match = f
                break
        if match is None:
            raise FileNotFoundError(f"codex session not found: {id}")
        messages: list[SessionMessage] = []
        for entry in read_jsonl(match):
            payload = entry.get("payload") or {}
            role = entry.get("role") or payload.get("role")
            content = entry.get("content") or payload.get("content")
            if not role or content is None:
                continue
            messages.append(
                SessionMessage(
                    role=role,
                    content=_flatten(content),
                    timestamp=entry.get("timestamp"),
                    raw=entry,
                )
            )
        t = file_times(match)
        return SessionDetail(
            id=id,
            provider="codex",
            path=str(match),
            messageCount=len(messages),
            createdAt=t["createdAt"],
            updatedAt=t["updatedAt"],
            messages=messages,
        )

    async def run(
        self,
        prompt: str,
        *,
        session_id: str | None = None,
        cwd: str | None = None,
        yolo: bool | None = None,
    ) -> RunResult:
        from codex_app_server_sdk import CodexClient

        is_yolo = yolo_default() if yolo is None else yolo
        config = _build_thread_config(is_yolo, cwd)

        async with CodexClient.connect_stdio(command=_codex_command()) as client:
            await client.initialize()
            if session_id:
                thread = await client.resume_thread(session_id, overrides=config)
            else:
                thread = await client.start_thread(config=config)
            result = await client.chat_once(prompt, thread_id=thread.thread_id)

        # ChatResult exposes assistant text + items; pick the most useful field.
        output = ""
        for attr in ("assistant_text", "final_response", "text"):
            v = getattr(result, attr, None)
            if v:
                output = v
                break
        return RunResult(
            sessionId=getattr(thread, "thread_id", None) or session_id,
            output=output,
            raw=result,
        )

File: src/providers/codex.ts
import { homedir } from "node:os";
import { join, basename } from "node:path";
import { existsSync } from "node:fs";
import fg from "fast-glob";
import { startRun } from "../runs/start.js";
import { planAiSessionResolution } from "../ai-sessions/finalize.js";
import { buildCatalog } from "../skills/catalog.js";
import { loadDotenv } from "../sessions/dotenv.js";
import type { RunHandle } from "../runs/types.js";
import type {
  Provider,
  SessionSummary,
  SessionDetail,
  SessionMessage,
  RunOptions,
} from "./types.js";
import { defaultYolo } from "./types.js";
import { readJsonl, fileTimes } from "../sessions/jsonl.js";
import { CodexAppServer } from "./codex-rpc.js";

const codexHome = () => process.env.CODEX_HOME || join(homedir(), ".codex");
const sessionsDir = () => join(codexHome(), "sessions");

interface CodexEntry {
  type?: string;
  role?: string;
  content?: unknown;
  timestamp?: string;
  payload?: { role?: string; content?: unknown };
}

function flattenContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((c: any) => (typeof c === "string" ? c : c?.text ?? JSON.stringify(c)))
      .join("\n");
  }
  return content == null ? "" : JSON.stringify(content);
}

function deriveId(path: string): string {
  const base = basename(path, ".jsonl");
  const idx = base.lastIndexOf("-");
  return idx > 0 ? base.slice(idx + 1) : base;
}

export const codexProvider: Provider = {
  name: "codex",

  async isAvailable() {
    return existsSync(sessionsDir());
  },

  async listSessions(): Promise<SessionSummary[]> {
    const dir = sessionsDir();
    if (!existsSync(dir)) return [];
    const files = await fg("**/*.jsonl", { cwd: dir, absolute: true });
    const out: SessionSummary[] = files.map((f) => {
      const t = fileTimes(f);
      return {
        id: deriveId(f),
        provider: "codex",
        path: f,
        createdAt: t.createdAt,
        updatedAt: t.updatedAt,
      };
    });
    return out.sort((a, b) => (b.updatedAt ?? "").localeCompare(a.updatedAt ?? ""));
  },

  async getSession(id: string): Promise<SessionDetail> {
    const dir = sessionsDir();
    const all = await fg("**/*.jsonl", { cwd: dir, absolute: true });
    const path = all.find((p) => deriveId(p) === id || basename(p, ".jsonl") === id);
    if (!path) throw new Error(`codex session not found: ${id}`);
    const entries = await readJsonl<CodexEntry>(path);
    const messages: SessionMessage[] = [];
    for (const e of entries) {
      const role = e.role ?? e.payload?.role;
      const content = e.content ?? e.payload?.content;
      if (!role || content == null) continue;
      messages.push({
        role: role as SessionMessage["role"],
        content: flattenContent(content),
        timestamp: e.timestamp,
        raw: e,
      });
    }
    const t = fileTimes(path);
    return {
      id,
      provider: "codex",
      path,
      messageCount: messages.length,
      createdAt: t.createdAt,
      updatedAt: t.updatedAt,
      messages,
    };
  },

  // Talks JSON-RPC NDJSON to `codex app-server` directly. Unlocks steer +
  // clean interrupt that the npm SDK doesn't expose.
  run(opts: RunOptions): RunHandle {
    const yolo = opts.yolo ?? defaultYolo();
    const plan = planAiSessionResolution({
      provider: "codex",
      prompt: opts.prompt,
      sessionId: opts.sessionId,
      asId: opts.aiSessionId,
      cwd: opts.cwd,
      internal: opts.internal,
    });
    const effectiveSessionId = plan.effectiveProviderSessionId;
    const effectiveCwd = plan.effectiveCwd;
    return startRun({
      provider: "codex",
      prompt: opts.prompt,
      sessionId: effectiveSessionId,
      cwd: effectiveCwd,
      yolo,
      internal: opts.internal,
      aiSessionId: plan.preResolvedAiSessionId,
      onFinalize: plan.attachToMeta,
      steerable: true,
      body: async ({ emit, onAbort, onSteer }) => {
        const workspaceEnv = loadDotenv(effectiveCwd);
        const client = new CodexAppServer({
          cwd: effectiveCwd,
          env: { ...process.env, ...workspaceEnv },
        });
        try {
          await client.request("initialize", {
            clientInfo: {
              name: "ai-sessions",
              title: "ai-sessions",
              version: "0.1.0",
            },
          });

          const threadParams: Record<string, unknown> = {};
          if (effectiveCwd) threadParams.cwd = effectiveCwd;
          if (yolo) {
            threadParams.sandbox = "danger-full-access";
            threadParams.approvalPolicy = "never";
          }
          const skillsCatalog = effectiveCwd ? buildCatalog(effectiveCwd) : "";
          if (skillsCatalog) {
            threadParams.developer_instructions = skillsCatalog;
          }

          const threadResult: any = effectiveSessionId
            ? await client.request("thread/resume", {
                threadId: effectiveSessionId,
                ...threadParams,
              })
            : await client.request("thread/start", threadParams);

          const threadId: string =
            threadResult?.thread?.id ?? effectiveSessionId ?? "";
          if (threadId) emit({ type: "session_id", sessionId: threadId });

          // State across notifications for this turn.
          let turnId: string | null = null;
          let textOut = "";
          const turnDone = new Promise<{ status: string; error?: string }>((resolve) => {
            const unsubs: Array<() => void> = [];
            const cleanup = () => unsubs.forEach((fn) => fn());

            unsubs.push(
              client.on("turn/started", (p: any) => {
                const tid = p?.turn?.id ?? p?.turnId;
                if (tid && !turnId) turnId = tid;
              })
            );
            unsubs.push(
              client.on("item/agentMessage/delta", (p: any) => {
                if (p?.delta) {
                  textOut += p.delta;
                  emit({ type: "text", text: p.delta });
                }
              })
            );
            unsubs.push(
              client.on("item/completed", (p: any) => {
                const item = p?.item;
                if (!item) return;
                if (item.type === "agent_message" && item.text) {
                  if (!textOut) {
                    textOut = item.text;
                    emit({ type: "text", text: item.text });
                  }
                } else if (item.type === "command_execution") {
                  emit({
                    type: "tool_use",
                    name: "command_execution",
                    input: { command: item.command, status: item.status },
                  });
                } else if (item.type === "error") {
                  emit({ type: "error", message: item.message ?? "codex error" });
                }
              })
            );
            unsubs.push(
              client.on("error", (p: any) => {
                const msg = p?.error?.message;
                if (msg) emit({ type: "error", message: msg });
              })
            );
            unsubs.push(
              client.on("turn/completed", (p: any) => {
                cleanup();
                const status = p?.turn?.status ?? "completed";
                const error = p?.turn?.error?.message;
                resolve({ status, ...(error ? { error } : {}) });
              })
            );
            unsubs.push(
              client.on("turn/failed", (p: any) => {
                cleanup();
                resolve({
                  status: "failed",
                  error: p?.error?.message ?? "turn failed",
                });
              })
            );
          });

          // Build input items: text + any image attachments as local_image.
          // Document attachments get appended to the text as path references.
          const inputItems: any[] = [];
          const docPaths = (opts.attachments ?? [])
            .filter((a) => a.kind === "document")
            .map((a) => a.path);
          const textBody =
            opts.prompt +
            (docPaths.length
              ? "\n" + docPaths.map((p) => `[Attached file: ${p}]`).join("\n")
              : "");
          if (textBody) inputItems.push({ type: "text", text: textBody });
          for (const a of opts.attachments ?? []) {
            if (a.kind === "image") inputItems.push({ type: "local_image", path: a.path });
          }
          // Capture turnId from the response too (for steer/interrupt).
          const turnStartResultPromise = client
            .request("turn/start", {
              threadId,
              input: inputItems,
            })
            .then((res: any) => {
              if (res?.turn?.id && !turnId) turnId = res.turn.id;
              return res;
            })
            .catch((err: Error) => {
              emit({ type: "error", message: err.message });
              throw err;
            });

          onAbort(async () => {
            if (threadId && turnId) {
              try {
                await client.request("turn/interrupt", { threadId, turnId });
              } catch {
                /* best effort */
              }
            }
          });

          onSteer?.(async (input: string) => {
            if (!threadId || !turnId) return;
            try {
              await client.request("turn/steer", {
                threadId,
                expectedTurnId: turnId,
                input: [{ type: "text", text: input }],
              });
            } catch (e) {
              emit({
                type: "error",
                message: `steer failed: ${e instanceof Error ? e.message : String(e)}`,
              });
            }
          });

          await turnStartResultPromise;
          const final = await turnDone;
          // Top-level `error` notification already emitted the error event;
          // here we just propagate failure to the run framework.
          if (final.status === "failed") {
            throw new Error(final.error ?? "codex turn failed");
          }
          return { output: textOut, sessionId: threadId || undefined };
        } finally {
          await client.close();
        }
      },
    });
  },
};
